### DISCUSSION TITLE
JDK 17 - strange profiling results

### DISCUSSION THREAD

**Discussion by krzysztofslusarski** [async-profiler/async-profiler#541] [Q&A] [answered]
Hi, I have a feeling that profiling application on JDK 17 (Amazon Coretto dist) gives very missleading output. For example I found such a stacktrace in my application:
![image](https://user-images.githubusercontent.com/57556371/151943821-ff8c5785-d9b4-4151-b5d5-dffbc469d261.png)
But I'm completely sure that the ```ArrayList``` has a proper size, and it is not growing while adding. 

When I profile the same application on JDK 11 there is no such a strange output, in example above there is no ```grow```.

The ```ArrayList``` is just an example, I found many strange results on JDK 17 that are not visible on JDK 11.

@apangin do you know was there any change in JVM that could create such a strange behaviour?

**Comment by apangin** [comment]
Interesting. I haven't observed such behavior, but on the other hand, I don't have have much history of profiling JDK 17; we just start using JDK 17 in production - probably we'll face the issue some time soon.

As usual, a test case would be extremely helpful.

**Comment by krzysztofslusarski** [comment]
I managed to reproduce the problem, I hope it will work on your machine too. I couldn't reproduce it with some basic java class with simple ```main``` method, but with Spring Boot I could:

```java
class TempBean {
    private static final int SIZE = 2048;
    private volatile List<Object> array;

    @Transactional
    public void execute() {
        List<Object> array = new ArrayList<>(SIZE);
        add(array);
        this.array = array;
    }

    private void add(List<Object> list) {
        addIt((integer1, integer2) -> list.add(new AbstractMap.SimpleEntry<>(integer1, integer2)));
    }

    private void addIt(BiConsumer<Integer, Integer> entryConsumer) {
        IntStream.range(0, SIZE).forEach(i -> entryConsumer.accept(i, i));
    }
}
```

I compiled the project with JDK 1.8:
```
openjdk version "1.8.0_312"
OpenJDK Runtime Environment Corretto-8.312.07.1 (build 1.8.0_312-b07)
OpenJDK 64-Bit Server VM Corretto-8.312.07.1 (build 25.312-b07, mixed mode)
```

And run it with JDK 17:
```
openjdk version "17.0.2" 2022-01-18 LTS
OpenJDK Runtime Environment Corretto-17.0.2.8.1 (build 17.0.2+8-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.2.8.1 (build 17.0.2+8-LTS, mixed mode, sharing)
```
Sources: [sources.zip](https://github.com/jvm-profiling-tools/async-profiler/files/7992679/sources.zip)
Binary: https://drive.google.com/file/d/1L1PYBwSo0ZCu2aXuKHoIbbcbHVJwv1Hg/view?usp=sharing

Running the JVM:
```shell
~/.jdks/corretto-17.0.2/bin/java -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints -agentpath:/home/pasq/Apps/AsyncProfiler-2.6/build/libasyncProfiler.so=start,event=cpu,lock,file=/tmp/boot/cpulock.jfr -jar Temp-0.0.1-SNAPSHOT.jar
```

Running the code on that JVM
```shell
for i in `seq 1 20000`; do echo $i; curl http://localhost:8080/; done
```

JDK 17 result:
![image](https://user-images.githubusercontent.com/57556371/152302410-7d3faf61-a649-4768-8adb-cb4b4f8e3607.png)
[jdk17.zip](https://github.com/jvm-profiling-tools/async-profiler/files/7992749/jdk17.zip)

JDK 8 result:
![image](https://user-images.githubusercontent.com/57556371/152303946-279bf042-5fad-44e4-8f91-6a31d7128e20.png)
It doesn't look great on flame graph, but there is no ```grow()``` method, just ```ensureCapacityInternal```
[jdk8.zip](https://github.com/jvm-profiling-tools/async-profiler/files/7992796/jdk8.zip)

Enjoy :)

**Comment by krzysztofslusarski** [comment, marked as answer]
JProfiler (async mode) has same problem, so I guess it might be an issue in JVM:

![image](https://user-images.githubusercontent.com/57556371/152355523-9e9f0feb-3d1c-4ce4-ad39-adc9c3a7f690.png)

**Reply by franz1981** [reply]
If you can try with `perf java map` it would help to check it!

(Sorry that I've added my comment here, but this issue capture my attention :) )

**Reply by toktarev** [reply]
I think that no need to say "sorry"
This is free space for people from community

**Reply by krzysztofslusarski** [reply]
Same story with perf:

```
http-nio-8080-e-?/635295;Lcom/example/temp/TempBean$$Lambda$804.0x0000000801202918:::accept;Lcom/example/temp/TempBean:::lambda$addIt$1_[i];Lcom/example/temp/TempBean$$Lambda$803.0x00000008012026e0:::accept_[i];Lcom/example/temp/TempBean:::lambda$add$0_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 1
http-nio-8080-e-?/635304;Lcom/example/temp/TempBean$$Lambda$804.0x0000000801202918:::accept;Lcom/example/temp/TempBean:::lambda$addIt$1_[i];Lcom/example/temp/TempBean$$Lambda$803.0x00000008012026e0:::accept_[i];Lcom/example/temp/TempBean:::lambda$add$0_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 7
http-nio-8080-e-?/635307;Lcom/example/temp/TempBean$$Lambda$804.0x0000000801202918:::accept;Lcom/example/temp/TempBean:::lambda$addIt$1_[i];Lcom/example/temp/TempBean$$Lambda$803.0x00000008012026e0:::accept_[i];Lcom/example/temp/TempBean:::lambda$add$0_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 3
http-nio-8080-e-?/635310;Lcom/example/temp/TempBean$$Lambda$804.0x0000000801202918:::accept;Lcom/example/temp/TempBean:::lambda$addIt$1_[i];Lcom/example/temp/TempBean$$Lambda$803.0x00000008012026e0:::accept_[i];Lcom/example/temp/TempBean:::lambda$add$0_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 4
http-nio-8080-e-?/635310;Lorg/springframework/web/context/request/async/WebAsyncManager:::setAsyncWebRequest;Lorg/springframework/web/context/request/async/StandardServletAsyncWebRequest:::addCompletionHandler_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 1
http-nio-8080-e-?/635311;Lcom/example/temp/TempBean$$Lambda$804.0x0000000801202918:::accept;Lcom/example/temp/TempBean:::lambda$addIt$1_[i];Lcom/example/temp/TempBean$$Lambda$803.0x00000008012026e0:::accept_[i];Lcom/example/temp/TempBean:::lambda$add$0_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 1
http-nio-8080-e-?/635314;Lcom/example/temp/TempBean$$Lambda$804.0x0000000801202918:::accept;Lcom/example/temp/TempBean:::lambda$addIt$1_[i];Lcom/example/temp/TempBean$$Lambda$803.0x00000008012026e0:::accept_[i];Lcom/example/temp/TempBean:::lambda$add$0_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 3
http-nio-8080-e-?/635314;Lorg/springframework/web/context/request/async/WebAsyncManager:::setAsyncWebRequest;Lorg/springframework/web/context/request/async/StandardServletAsyncWebRequest:::addCompletionHandler_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 1
http-nio-8080-e-?/635318;Lcom/example/temp/TempBean$$Lambda$804.0x0000000801202918:::accept;Lcom/example/temp/TempBean:::lambda$addIt$1_[i];Lcom/example/temp/TempBean$$Lambda$803.0x00000008012026e0:::accept_[i];Lcom/example/temp/TempBean:::lambda$add$0_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 2
http-nio-8080-e-?/635318;Lorg/springframework/web/context/request/async/WebAsyncManager:::setAsyncWebRequest;Lorg/springframework/web/context/request/async/StandardServletAsyncWebRequest:::addCompletionHandler_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::add_[i];Ljava/util/ArrayList:::grow_[i];Ljava/util/ArrayList:::grow_[i] 1
```

This is getting wierd, I thought it might be ```AsyncGetCallTrace``` issue, but perf doesn't use it.

**Reply by apangin** [reply]
So the problem is apparently in the debug information embedded in the JIT compiled code: machine instructions are attributed to the wrong bytecodes. I'll look into this when I have time.

**Reply by krzysztofslusarski** [reply]
As I said in the beginning, the ```ArrayList``` is just an example, but easy to recreate. I have multiple issues when the ouput doesn't look valid. Hope that recreation helps in finding the source of a problem.

**Reply by apangin** [reply]
I was able to reproduce the issue with a reduced single-class example:
```
import java.util.ArrayList;
import java.util.List;

public class ArrayListGrow {
   static final int SIZE = 2048;
   static volatile List<Object> tmp;

    static void runTest() {
        List<Object> list = new ArrayList<>(SIZE);
        fill(list);
        tmp = list;
    }

    static void fill(List<Object> list) {
        for (int i = 0; i < SIZE; i++) {
            list.add(i);
        }
    }

    static void spoil() {
        for (int i = 0; i < 1000000; i++) {
            new ArrayList<>(0).add("");
        }
    }

    public static void main(String[] args) {
        spoil();

        while (true) {
            runTest();
        }
    }
}
```

Indeed, instruction-to-bytecode mapping appears wrong, so many tools show misleading information, including the original Flight Recorder and JMH.

The bug is not specific to JDK 17 - I see the same behavior on 11.0.11.

Interestingly, if I comment out `// spoil()` call, the profile will be correct.

I'll check if there is an existing JDK bug, and if not, will submit a new one.

**Reply by krzysztofslusarski** [reply]
Thank you @apangin.

**Reply by krzysztofslusarski** [reply]
@apangin Can you please paste the JDK bug number here so we could track the fix version?

**Reply by apangin** [reply]
I've submitted https://bugs.openjdk.java.net/browse/JDK-8281677

**Reply by krzysztofslusarski** [reply]
Thank you
